File: connectors/cpp/tests/integration/core/c10_request_timeout_test.cpp
#include "../base_integration_test.hpp"
#include <gtest/gtest.h>

using namespace playhouse;
using namespace playhouse::test;

/// C-10: Request Timeout Tests
/// Verifies request timeout handling
class C10_RequestTimeoutTest : public BaseIntegrationTest {};

TEST_F(C10_RequestTimeoutTest, Request_WithShortTimeout_TimesOut) {
    // Given: Connected with very short timeout
    config_.request_timeout_ms = 100;  // 100ms timeout
    connector_ = std::make_unique<Connector>();
    connector_->Init(config_);

    ASSERT_TRUE(CreateStageConnectAndAuthenticate("timeout_short_user"));

    // When: Send a request that won't respond in time
    Bytes payload = proto::EncodeNoResponseRequest(5000);
    auto packet = Packet::FromBytes("NoResponseRequest", std::move(payload));

    Packet response = Packet::Empty("Empty");
    bool completed = RequestAndWait(std::move(packet), response, 500);

    // Then: Should timeout or return timeout message
    bool timed_out = !completed;
    if (completed) {
        if (response.GetMsgId() == msg_id::TIMEOUT || response.GetErrorCode() != 0) {
            timed_out = true;
        }
    }

    EXPECT_TRUE(timed_out) << "Request should timeout with short timeout configuration";
}

TEST_F(C10_RequestTimeoutTest, Request_WithNormalTimeout_Succeeds) {
    // Given: Connected with normal timeout
    config_.request_timeout_ms = 5000;  // 5 second timeout
    connector_ = std::make_unique<Connector>();
    connector_->Init(config_);

    ASSERT_TRUE(CreateStageConnectAndAuthenticate("timeout_normal_user"));

    // When: Send a normal request
    Bytes payload = proto::EncodeEchoRequest("Normal request", 1);
    auto packet = Packet::FromBytes("EchoRequest", std::move(payload));

    Packet response = Packet::Empty("Empty");
    bool completed = RequestAndWait(std::move(packet), response, 6000);

    // Then: Should succeed before timeout
    ASSERT_TRUE(completed) << "Normal request should complete";
    EXPECT_EQ(response.GetErrorCode(), 0) << "Normal request should succeed";
}

TEST_F(C10_RequestTimeoutTest, Request_TimeoutMessage_HasTimeoutId) {
    // Given: Connected with short timeout
    config_.request_timeout_ms = 200;
    connector_ = std::make_unique<Connector>();
    connector_->Init(config_);

    ASSERT_TRUE(CreateStageConnectAndAuthenticate("timeout_msgid_user"));

    // When: Send request that will timeout
    Bytes payload = proto::EncodeNoResponseRequest(10000);
    auto packet = Packet::FromBytes("NoResponseRequest", std::move(payload));

    Packet response = Packet::Empty("Empty");
    bool completed = RequestAndWait(std::move(packet), response, 500);

    // Then: Timeout message should have special ID or no response
    if (!completed) {
        SUCCEED() << "Request timed out without response";
        return;
    }

    EXPECT_EQ(response.GetMsgId(), msg_id::TIMEOUT);
}

TEST_F(C10_RequestTimeoutTest, Request_MultipleTimeouts_AllHandledCorrectly) {
    // Given: Connected with short timeout
    config_.request_timeout_ms = 150;
    connector_ = std::make_unique<Connector>();
    connector_->Init(config_);

    ASSERT_TRUE(CreateStageConnectAndAuthenticate("timeout_multi_user"));

    // When: Send multiple requests that will timeout
    int timeout_count = 0;
    for (int i = 0; i < 3; i++) {
        Bytes payload = proto::EncodeNoResponseRequest(5000);
        auto packet = Packet::FromBytes("NoResponseRequest", std::move(payload));

        Packet response = Packet::Empty("Empty");
        bool completed = RequestAndWait(std::move(packet), response, 200);
        if (!completed) {
            timeout_count++;
        } else if (response.GetMsgId() == msg_id::TIMEOUT || response.GetErrorCode() != 0) {
            timeout_count++;
        }
    }

    // Then: All should timeout
    EXPECT_GT(timeout_count, 0) << "At least some requests should timeout";
}

TEST_F(C10_RequestTimeoutTest, Request_AfterTimeout_CanStillSendNew) {
    // Given: Connected and had a timeout
    config_.request_timeout_ms = 100;
    connector_ = std::make_unique<Connector>();
    connector_->Init(config_);

    ASSERT_TRUE(CreateStageConnectAndAuthenticate("timeout_after_user"));

    // First request that times out
    Bytes slow_payload = proto::EncodeNoResponseRequest(5000);
    auto slow_packet = Packet::FromBytes("NoResponseRequest", std::move(slow_payload));
    Packet slow_response = Packet::Empty("Empty");
    RequestAndWait(std::move(slow_packet), slow_response, 200);

    // When: Send a new fast request
    Bytes fast_payload = proto::EncodeEchoRequest("Fast request", 1);
    auto fast_packet = Packet::FromBytes("EchoRequest", std::move(fast_payload));
    Packet fast_response = Packet::Empty("Empty");
    bool completed = RequestAndWait(std::move(fast_packet), fast_response, 5000);

    // Then: New request should still work
    if (completed) {
        EXPECT_TRUE(connector_->IsConnected()) << "Connection should still be valid after timeout";
    } else {
        SUCCEED() << "System stable after timeout";
    }
}

File: connectors/javascript/tests/integration/core/C10_RequestTimeout.test.ts
/**
 * C-10: Request Timeout Tests
 *
 * Tests that requests timeout when the server does not respond.
 * NoResponseRequest causes the server to intentionally not respond.
 */

import { describe, test, expect, beforeEach, afterEach } from 'vitest';
import { BaseIntegrationTest } from '../helpers/BaseIntegrationTest.js';
import { Connector } from '../../../src/connector.js';
import { Packet } from '../../../src/packet.js';
import { ErrorCode } from '../../../src/types.js';
import { serializeNoResponseRequest } from '../helpers/TestMessages.js';

describe('C-10: Request Timeout', () => {
    let testContext: BaseIntegrationTest;

    beforeEach(async () => {
        testContext = new BaseIntegrationTest();
        await testContext['beforeEach']();
    });

    afterEach(async () => {
        await testContext['afterEach']();
    });

    test('C-10-01: Request without response times out', async () => {
        // Given: Short timeout setting (2 seconds)
        const connector = new Connector();
        connector.init({
            requestTimeoutMs: 2000, // 2 second timeout
            heartbeatIntervalMs: 10000
        });
        testContext['connector'] = connector;

        await testContext['createStageAndConnect']();
        await testContext['authenticate']('timeoutUser');

        const noResponseRequest = {
            delayMs: 10000 // Server won't respond for 10 seconds
        };

        // When: Send request without response
        const payload = serializeNoResponseRequest(noResponseRequest);
        const requestPacket = Packet.fromBytes('NoResponseRequest', payload);

        // Then: Should timeout
        await expect(
            connector.request(requestPacket)
        ).rejects.toThrow();
    });

    test('C-10-02: Connection remains after timeout', async () => {
        // Given: Short timeout setting
        const connector = new Connector();
        connector.init({
            requestTimeoutMs: 2000,
            heartbeatIntervalMs: 10000
        });
        testContext['connector'] = connector;

        await testContext['createStageAndConnect']();
        await testContext['authenticate']('timeoutConnectionUser');

        const noResponseRequest = {
            delayMs: 10000
        };

        // When: Timeout occurs
        const payload = serializeNoResponseRequest(noResponseRequest);
        const requestPacket = Packet.fromBytes('NoResponseRequest', payload);
        try {
            await connector.request(requestPacket);
        } catch (error) {
            // Ignore timeout exception
        }

        // Then: Connection should remain
        expect(connector.isConnected).toBe(true);
        expect(connector.isAuthenticated).toBe(true);

        // Other requests should work normally
        const echoReply = await testContext['echo']('After Timeout', 1);
        expect(echoReply.content).toBe('After Timeout');
    });

    test('C-10-03: Callback-based request also times out', async () => {
        // Given: Short timeout setting
        const connector = new Connector();
        connector.init({
            requestTimeoutMs: 2000,
            heartbeatIntervalMs: 10000
        });
        testContext['connector'] = connector;

        await testContext['createStageAndConnect']();
        await testContext['authenticate']('callbackTimeoutUser');

        const noResponseRequest = {
            delayMs: 10000
        };

        let callbackInvoked = false;
        let errorCode: number | undefined;
        let errorOccurred = false;

        connector.onError = (_stageId, code, _message) => {
            errorCode = code;
            errorOccurred = true;
        };

        // When: Callback-based request without response
        const payload = serializeNoResponseRequest(noResponseRequest);
        const requestPacket = Packet.fromBytes('NoResponseRequest', payload);
        connector.requestWithCallback(requestPacket, () => {
            callbackInvoked = true;
        });

        // Wait for OnError event (with mainThreadAction, max 5 seconds)
        const completed = await testContext['waitForConditionWithMainThread'](
            () => errorOccurred,
            5000
        );

        // Then: OnError event should occur and callback should not be invoked
        expect(completed).toBe(true);
        expect(errorCode).toBe(ErrorCode.RequestTimeout);
        expect(callbackInvoked).toBe(false);
    });

    test('C-10-04: One timeout among multiple requests does not affect others', async () => {
        // Given: Short timeout setting
        const connector = new Connector();
        connector.init({
            requestTimeoutMs: 2000,
            heartbeatIntervalMs: 10000
        });
        testContext['connector'] = connector;

        await testContext['createStageAndConnect']();
        await testContext['authenticate']('multiTimeoutUser');

        // When: Send normal and timeout requests in parallel
        const echoTask1 = testContext['echo']('Normal 1', 1);
        const echoTask2 = testContext['echo']('Normal 2', 2);

        const noResponseRequest = {
            delayMs: 10000
        };
        const payload = serializeNoResponseRequest(noResponseRequest);
        const timeoutPacket = Packet.fromBytes('NoResponseRequest', payload);
        const timeoutTask = connector.request(timeoutPacket);

        const echoTask3 = testContext['echo']('Normal 3', 3);

        // Wait for normal requests to complete
        const echo1 = await echoTask1;
        const echo2 = await echoTask2;
        const echo3 = await echoTask3;

        // Then: Normal requests should succeed
        expect(echo1.content).toBe('Normal 1');
        expect(echo2.content).toBe('Normal 2');
        expect(echo3.content).toBe('Normal 3');

        // Timeout request should fail
        await expect(timeoutTask).rejects.toThrow();
    });

    test('C-10-05: Long timeout setting allows response to be received', async () => {
        // Given: Long timeout setting (15 seconds)
        const connector = new Connector();
        connector.init({
            requestTimeoutMs: 15000, // 15 second timeout
            heartbeatIntervalMs: 10000
        });
        testContext['connector'] = connector;

        await testContext['createStageAndConnect']();
        await testContext['authenticate']('longTimeoutUser');

        // When: Send echo request with short delay (response within timeout)
        const echoReply = await testContext['echo']('Long Timeout Test', 1);

        // Then: Should receive response normally
        expect(echoReply.content).toBe('Long Timeout Test');
    });

    test('C-10-06: Authentication request can also timeout', async () => {
        // Given: Very short timeout setting (100ms)
        const connector = new Connector();
        connector.init({
            requestTimeoutMs: 100, // 100ms timeout
            heartbeatIntervalMs: 10000
        });
        testContext['connector'] = connector;

        await testContext['createStageAndConnect']();

        // When: Authentication request (may timeout with network delay)
        // Note: This test may be unstable depending on network conditions
        // Timeout may not actually occur
        const exception = await connector.authenticate('TestStage', 'authTimeoutUser')
            .then(() => null)
            .catch((error) => error);

        // Then: If exception occurs, it should be about timeout or connection
        if (exception) {
            expect(exception.message).toBeDefined();
        }
    });

    test('C-10-07: Timeout exception contains request information', async () => {
        // Given: Short timeout setting
        const connector = new Connector();
        connector.init({
            requestTimeoutMs: 2000,
            heartbeatIntervalMs: 10000
        });
        testContext['connector'] = connector;

        await testContext['createStageAndConnect']();
        await testContext['authenticate']('exceptionInfoUser');

        const noResponseRequest = {
            delayMs: 10000
        };

        // When: Timeout occurs
        const payload = serializeNoResponseRequest(noResponseRequest);
        const requestPacket = Packet.fromBytes('NoResponseRequest', payload);
        let caughtException: Error | null = null;

        try {
            await connector.request(requestPacket);
        } catch (error) {
            caughtException = error as Error;
        }

        // Then: Exception should contain request information
        expect(caughtException).not.toBeNull();
        expect(caughtException!.message).toBeDefined();
        expect(caughtException!.message).toContain('timeout');
    });
});
